src/coreclr/jit/flowgraph.cpp

```diff
@@ -4619,6 +4621,181 @@ bool FlowGraphNaturalLoops::FindNaturalLoopBlocks(FlowGraphNaturalLoop* loop, Ar
     return true;
 }
 
+#ifdef DEBUG
+
+//------------------------------------------------------------------------
+// FlowGraphNaturalLoop::Dump: print one loops to the JitDump
+//
+/* static */
+void FlowGraphNaturalLoop::Dump(FlowGraphNaturalLoop* loop)
+{
+    if (loop == nullptr)
+    {
+        printf("loop is nullptr");
+        return;
+    }
+
+    BitVecTraits loopTraits(loop->GetDfsTree()->PostOrderTraits());
+    BitVec       outputBlocks(BitVecOps::MakeEmpty(&loopTraits));
+
+    // Display: LOOP# (old LOOP#) / header / parent loop# / blocks / entry edges / exit edges / back edges
+    // Blocks can compacted be "[top .. bottom]" if lexically adjacent and no non-loop blocks in
+    // the range. Otherwise, print a verbose list of blocks.
+
+    Compiler* comp = loop->GetDfsTree()->GetCompiler();
+    printf(FMT_LP, loop->GetIndex());
+
+    // We might want to print out the old loop number using something like:
+    //
+    // if (comp->m_newToOldLoop[loop->GetIndex()] != nullptr)
+    // {
+    //   printf(" (old: " FMT_LP ")", (unsigned)(comp->m_newToOldLoop[loop->GetIndex()] - comp->optLoopTable));
+    // }
+    //
+    // However, not all callers of FlowGraphNaturalLoops::Find update m_newToOldLoop -- only
+    // Compiler::optFindNewLoops() does that. This dumper should work with any construction of
+    // FlowGraphNaturalLoops.
+
+    printf(" header: " FMT_BB, loop->GetHeader()->bbNum);
+    if (loop->GetParent() != nullptr)
+    {
+        printf(" parent: " FMT_LP, loop->GetParent()->GetIndex());
+    }
+
+    // Dump the set of blocks in the loop. There are three cases:
+    // 1. If there is only one block in the loop, display it.
+    // 2. If the blocks happen to be lexically dense and without non-loop blocks in the range,
+    //    then use a shortcut of `[BBtop .. BBbottom]`.
+    // 3. Otherwise, display the entire list of blocks.
+    //
+    // Lexicality depends on properly renumbered blocks, which we might not have when dumping.
+
+    printf("\n  Members (%u): ", loop->NumLoopBlocks());
+
+    const unsigned numBlocks = loop->NumLoopBlocks();
+    if (numBlocks == 0)
+    {
+        // This should never happen?
+        printf("NONE?");
+    }
```

Comment:
Indeed, this can be an assert.

Reply:
Asserts in diagnostic helpers are annoying, so I'd keep this as is.

Reply:
Yes, dumpers should not have asserts, but should be written to be (reasonably) tolerant of buggy data structures (presumably there would be an equivalent assert someplace else).